src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -21,49 +19,56 @@
 using Microsoft.CodeAnalysis.Shared.Utilities;
 using Microsoft.CodeAnalysis.SymbolSearch;
 using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Utilities;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.AddMissingImports
 {
     internal abstract class AbstractAddMissingImportsFeatureService : IAddMissingImportsFeatureService
     {
         protected abstract ImmutableArray<string> FixableDiagnosticIds { get; }
 
-        public async Task<bool> HasMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        /// <inheritdoc/>
+        public async Task<Document?> AddMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
         {
-            // Get the diagnostics that indicate a missing import.
-            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
-            if (diagnostics.IsEmpty)
+            var analysisResult = await AnalyzeAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return await AddMissingImportsAsync(analysisResult, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc/>
+        public async Task<Document?> AddMissingImportsAsync(AddMissingImportsAnalysisResult analysisResult, CancellationToken cancellationToken)
+        {
+            if (analysisResult.CanAddMissingImports)
             {
-                return false;
+                // Apply those fixes to the document.
+                var newDocument = await ApplyFixesAsync(analysisResult.Document, analysisResult.AddImportFixData, cancellationToken).ConfigureAwait(false);
+                return newDocument;
             }
 
-            // Find fixes for the diagnostic where there is only a single fix.
-            var usableFixes = await GetUnambiguousFixesAsync(document, diagnostics, cancellationToken).ConfigureAwait(false);
-            return !usableFixes.IsEmpty;
+            return null;
```

Comment:
this feels weird to me.  I would think this entire API would just return the doc as-is if nothing changed.  i.e. the request is to add imports, and the service responds back with: here's the doc with the imports added.  even if that's no imports, here it is.

Reply:
There doesn't seem to be consistency on our internal APIs for this. Is there something documented somewhere? If not, it just seems like preference. I like knowing if something changed, especially since we work with immutable objects. Returning null is a great way to say that nothing happened, and is easier to use than `Try*` with a bool and out parameter.

I'll change this, but I don't see objective gain. If this is our standard we should document it. Regardless of my opinions  complying to the codebase norm is more important. We don't have that though :/